posts/20241209full-package-reviews/index.qmd

```diff
@@ -0,0 +1,57 @@
+
+Our regular code reviews take place on GitHub, using Pull Requests. For full package reviews, we also use Pull Requests in Epiverse-TRACE. This means that all the features from GitHub are available to make suggestions, approve the code, or leave precise comments at a specific line in the code.
+
+What is different, is that all the files are considered new, and part of the review. This includes the files already present in our [package template](https://github.com/epiverse-trace/packagetemplate), but also all test files, documentation, and more.
+
+## How do I set one up?
+
+Git and GitHub do not support full package reviews by default. This is because pull requests are usually merged into an existing branch, which already contains existing code. For a full package review, what we want is to merge onto an empty branch.
+
+In order to create a branch that has no code at all, run the following shell command from within your git directory:
+
+```sh
+git branch empty `git rev-list --all | tail -1`
+git push origin empty
+
+git branch review
+git push origin review
+```
+
+These commands are also noted in the blueprints, but I simplified them so that you do not need to copy paste the hash of the first commit. Also, you can use this command for any Git repository and open up the pull request through GitHub (or Gitlab, Forgejo). This means that doing a full package review can be done anywhere, anytime.
+
+## The experience of a full package review
+
+[Since joining Epiverse-TRACE](../chjh-intro-blog/index.qmd), I participated in many full package reviews, but did not until quite recently initiate my own. For the new package **safeframe**, [I recently requested a full package review](https://github.com/epiverse-trace/safeframe/pull/40). This highlighted two quirks I did not realize before.
+
+First, it feels like doing something that can easily go wrong. I have been using Git for over a decade[^1], but this was completely new to me. Given that git operations can sometimes cause issues, I was hesitant to just run the shell commands. Maybe having a demo video showing it works, will help contributors gain confidence in doing this.
+
+[^1]: My GitHub account is from 2011, but at that time I did not understand what GitHub was or did.
+
+Second, reviewer suggestions are great, but getting them on a full package review involves manual work to include them. I try to leave actionable suggestions directly in the code when I review, so they are easy to accept and integrate into the codebase. With full package reviews, receiving suggestions became slightly annoying. Not because of the reviewer, but because the suggestions have to be manually recreated in a separate branch that will actually be merged. It would be great if we had a way to keep the ease of integrating reviewer suggestions in full package reviews.
```

Comment:
This is a good point, it might be worth discussing if there is a way to alleviate some of this by how we set up code reviews. For example, we could see if it is possible to create PRs that target certain sections of the code that we want reviewing. 

Often in full package or partial package reviews, even actionable suggestions cannot be committed as I work on orthogonal branches which I then make the same change as suggested and add a co-author to the commit, and the review PR is closed without merging.

It would be great if we could offer some potential solutions within this blogpost, but it's definitely not a blocker to getting it published.